QccPack-0.61-1/src/libQccPackVID/utils/man1/memc.1
.TH MEMC 1 "QCCPACK" ""
.SH NAME
memc \- 
full-search, block-based motion estimation and compensation
.SH SYNOPSIS
.B memc 
.RB "[\|" \-fp "\|]"
.RB "[\|" \-hp "\|]"
.RB "[\|" \-qp "\|]"
.RB "[\|" \-ep "\|]"
.RB "[\|" \-f1 
.IR filter1 "\|]"
.RB "[\|" \-f2 
.IR filter2 "\|]"
.RB "[\|" \-f3 
.IR filter3 "\|]"
.RB "[\|" \-w
.IR window "\|]" 
.RB "[\|" \-b
.IR block_size "\|]" 
.RB "[\|" \-mc
.IR motion_compensated_frame "\|]" 
.RB "[\|" \-srf
.IR subpixel_reference_frame "\|]" 
.RB "[\|" \-r
.IR residual_frame "\|]" 
.I reference_frame current_frame mvfile
.SH DESCRIPTION
.TP
.B \-fp
Full-pixel accuracy.
.TP
.B \-hp
Half-pixel accuracy.
.TP
.B \-qp
Quarter-pixel accuracy.
.TP
.B \-ep
Eighth-pixel accuracy.
.TP
.BI \-f1 " filter1"
String. Filename of filter to use
for going from full-pixel to half-pixel accuracy.
.TP
.BI \-f2 " filter2"
String. Filename of filter to use
for going from half-pixel to quarter-pixel accuracy.
.TP
.BI \-f3 " filter3"
String. Filename of filter to use
for going from quarter-pixel to eighth-pixel accuracy.
.TP
.BI \-w " window"
Integer. Window size for motion estimation. Default is 15 pixels.
.TP
.BI \-b " block_size"
Integer. Block size. Default is 16 pixels.
.TP
.BI \-mc " motion_compensated_frame"
String. Filename of file to which to write the motion-compensated frame.
.TP
.BI \-srf " subpixel_reference_frame"
String. Filename of file to which to write the reference frame after
having been interpolated to the specified subpixel accuracy.
.TP
.BI \-r " residual_frame"
String. Filename of file to which to write the difference between
the current frame and the motion-compensated frame, the residual image
from the motion-compensation process.
.SH DESCRIPTION
.B memc
reads two images, a
.I current_frame
and a
.IR reference_frame ,
performing full-search, block-based motion estimation to estimate
motion from
.I reference_frame
to
.IR current_frame .
That is, for each block in
.IR current_frame ,
every block in a window surrounding the current-block location
in
.IR reference_frame
is searched, and the best-matching block chosen.
The size of the blocks to use for matching is given by
.I block_size
(default is 16); and the size of the window in which to perform
the block-matching search is
.IR window
(default is 15).
.LP
.BR memc
first interpolates
.I reference_frame
to the specified subpixel accuracy as determined by the
.BR \-fp ,
.BR \-hp ,
.BR \-qp ,
or
.BR \-ep 
options. In the case that none of these options are specified,
.BR memc
defaults to using full-pixel accuracy.
.BR QccVIDMotionEstimationCreateReferenceFrame (3)
is used to perform the subpixel-accuracy interpolation, with
.IR filter1 ,
.IR filter2 ,
and
.IR filter3
provided to the latter routine if specified
.RI ( filter1 ,
.IR filter2 ,
and
.IR filter3
should be files whose format is described with
.BR QccFilterRead (3)).
Next,
.BR QccVIDMotionEstimationFullSearch (3)
is called to carry out the full-search, block-based motion estimation.
.BR QccVIDMotionVectorsWriteFile (3)
then writes the resulting motion vectors to
.IR mvfile .
.LP
If the
.BR \-mc
option is specified,
.BR QccVIDMotionEstimationCreateCompensatedFrame (3)
is called to generate the motion-compensated frame using the
motion vectors that resulted from the full-search, block-based
motion search. This compensated frame is then written to
.IR motion_compensated_frame .
.LP
If the
.BR \-srf
option is specified,
the interpolated, subpixel-accurate reference frame (as generated by
.BR QccVIDMotionEstimationCreateReferenceFrame (3))
is written to
.IR reference_frame .
.LP
If the
.BR \-r
option is specified,
the residual image between the current frame and the motion-compensated
frame is written to
.IR residual_frame .
This residual is not the true difference image, but rather the absolute
value of the difference image, in order to avoid negative-valued
pixels.
.SH "SEE ALSO"
.BR QccVIDMotionEstimationFullSearch (3),
.BR QccVIDMotionEstimationCreateReferenceFrame (3),
.BR QccVIDMotionEstimationCreateCompensatedFrame (3),
.BR QccVIDMotionVectorsWriteFile (3),
.BR QccFilterRead (3),
.BR QccPackVID (3),
.BR QccPackIMG (3),
.BR QccPack (3)

.SH AUTHOR
Copyright (C) 1997-2016  James E. Fowler
.\"  The programs herein are free software; you can redistribute them and/or
.\"  modify them under the terms of the GNU General Public License
.\"  as published by the Free Software Foundation; either version 2
.\"  of the License, or (at your option) any later version.
.\"  
.\"  These programs are distributed in the hope that they will be useful,
.\"  but WITHOUT ANY WARRANTY; without even the implied warranty of
.\"  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\"  GNU General Public License for more details.
.\"  
.\"  You should have received a copy of the GNU General Public License
.\"  along with these programs; if not, write to the Free Software
.\"  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.


QccPack-0.61-1/CHANGELOG
==============================================================================

                                  Q C C P A C K

                            Version 0.61, 01-apr-2016

                    Copyright (C) 1997-2016  James E. Fowler

------------------------------------------------------------------------------


CHANGES:

Version 0.61:
  01-apr-2016:  BETA release version 0.61
  30-mar-2016:  Updated Cygwin installation instructions
  20-jan-2016:  Updated copyright year range for 2016
  20-jan-2016:  Fixed zero-subband-size bug in tce3d.c

Version 0.60:
  25-nov-2015:  BETA release version 0.60
  08-aug-2015:  Added QccFileReadLine()
  07-aug-2015:  Added QccFileGetRealPath()
  14-jul-2015:  Added QccFileWriteMagicNumberVersion()
  
Version 0.59:
  18-jun-2015:  BETA release version 0.59
  15-jun-2015:  Added QccFileRewind()
  14-jun-2015:  Updated copyright year range for 2015
  14-jun-2015:  Made QccENTExponentialGolombEncodeSymbol() and
                  QccENTExponentialGolombDecodeSymbol() visible
  09-jan-2014:  Updated copyright year range for 2014
  20-jul-2013:  Added Data/Matlab to QCCPACK_DATA_DIRS in Imakefile
  20-jul-2013:  Updated copyright year range for 2013
  21-dec-2011:  Updated copyright year range for 2011
  21-dec-2011:  Minor fixes for warning-free compilation with gcc-4.6
  23-jun-2009:  Added better support for uniform quantizers to sqencode,
                  sqdecode
  11-mar-2009:  Added logo.sed

Version 0.58:
  13-feb-2009:  BETA release version 0.58
  05-feb-2009:  Added klttcedecode3d.1, klttcedecode3d_lossless.1, 
                  klttceencode3d.1, klttceencode3d_lossless.1, 
                  tcedecode3d_lossless.1, tceencode3d_lossless.1
  05-feb-2009:  Added QccWAVtce3DLosslessEncode.3,
                  QccWAVtce3DLosslessDecodeHeader.3, 
                  QccWAVtce3DLosslessDecode.3, 
                  QccWAVklttce3DEncode.3, 
                  QccWAVklttce3DDecodeHeader.3, 
                  QccWAVklttce3DDecode.3, 
                  QccWAVklttce3DLosslessEncode.3, 
                  QccWAVklttce3DLosslessDecodeHeader.3, 
                  QccWAVklttce3DLosslessDecode.3
  05-feb-2009:  Added klttce3d.c, klttceencode3d.c, klttcedecode3d.c
  04-feb-2009:  Changed QccHYPrkltTrain(), QccHYPrkltTransform(),
                  QccHYPrkltInverseTransform() to use integer volumes
  04-feb-2009:  Added klttce3d_lossless.c, klttceencode3d_lossless.c,
                  klttcedecode3d_lossless.c
  04-feb-2009:  Fixed zero temporal_num_scales problem in 3D packet transforms
  02-feb-2009:  Added tce3d_lossless.c,
                  tceencode3d_lossless, tcedecode3d_lossless
  29-jan-2009:  QccHYPrkltTrain.3, QccHYPrkltFactorization.3,
                  QccHYPrkltTransform.3, QccHYPrkltInverseTransform.3
  29-jan-2009:  Added QccHYPrklt.3
  29-jan-2009:  Added QccHYPrkltInitialize(), QccHYPrkltAlloc(),
                  QccHYPrkltFree(), QccHYPrkltTrain(),
                  QccHYPrkltFactorization()
  28-jan-2009:  Added QccHYPrkltTransform(), QccHYPrkltInverseTransform()
  28-jan-2009:  Added QccMatrixIdentity(), QccMatrixRowExchange(),
                  QccMatrixColExchange(), QccMatrixIdentity.3,
                  QccMatrixRowExchange.3, QccMatrixColExchange.3
  16-jan-2009:  Updated copyright year range for 2009

Version 0.57:
  30-jun-2008:  BETA release version 0.57
  09-may-2008:  Added QccTimeTic(), QccTimeToc(), QccTime.3
  06-may-2008:  Fixed spatial_num_levels/temporal_num_levels bug in
                  initialization in bisk3d.c, wdr3d.c
  06-may-2008:  Fixed zero number of levels in
                  QccWAVWaveletPacketDWT3D(),
                  QccWAVWaveletInversePacketDWT3D()
  26-apr-2008:  Added sam.c, icbsam.c, QccHYPImageCubeMeanSAM(),
                QccVectorAngle()
  17-apr-2008:  Added QccWAVtce3DEncode.3, tceencode3d.1
  16-apr-2008:  Added tce3d.c, tceencode3d.c, tcedecode3d.c
  16-apr-2008:  Added -bpv to rawtoicb.c, icbtoraw.c, rawdist3d.c
  16-apr-2008:  Added -bpp to rawtoicp.c, icptoraw.c, rawdist2d.c
  16-apr-2008:  Added bpv, bpp to QccHYPRaw(), QccHYPRawRead2D(),
                  QccHYPRawWrite2D(), QccHYPRawDist2D(), QccHYPRawRead3D(),
                  QccHYPRawWrite3D(), QccHYPRawDist3D(),

Version 0.56:
  14-mar-2008:  BETA release version 0.56
  14-mar-2008:  Added QccIMGImageComponentLBT.3,
                  QccIMGImageComponentInverseLBT.3
  13-mar-2008:  Added QccFastDCT, QccFastDCTInitialize, QccFastDCTCreate,
                  QccFastDCTFree, QccFastDCTForwardTransform1D,
                  QccFastDCTInverseTransform1D, QccFastDCTForwardTransform2D,
                  QccFastDCTInverseTransform2D
  13-mar-2008:  Added fast_dct.c
  13-mar-2008:  Added dcttceencode.1, dcttcedecode.1
  13-mar-2008:  Added QccWAVdcttceEncode.3, QccWAVdcttceDecodeHeader.3,
                  QccWAVdcttceDecode.3
  13-mar-2008:  Renamed image_lapped_filter.c to image_lbt.c
  24-feb-2008:  Fixed memory leaks in dcttce.c, image_lapped_filter.c
  17-jan-2008:  Added Haar.fbk
  09-jan-2008:  Added dcttce.c, dcttceencode.c, dcttcedecode.c
  09-jan-2008:  Added image_lapped_filter.c
  09-jan-2008:  Added QccMatrixInverse(), QccMatrixInverse.3
  04-jan-2008:  Updated substr
  04-jan-2008:  Updated copyright year range for 2008
  11-dec-2007:  Fixed minor max_coefficients bug in QccWAVtceEncodeDWT()
  24-oct-2007:  Changed spawn to be able to handle commands with arguments
  09-oct-2007:  Added QCCMATHEPS
  01-oct-2007:  Added QccMatrixIntMultiply(), QccMatrixIntMultiply.3
  01-oct-2007:  Added QccMatrixMultiply(), QccMatrixMultiply.3
  28-sep-2007:  Added QccMatrixNullspace(), QccMatrixNullspace.3
  28-sep-2007:  Added QccMatrixOrthogonalize(), QccMatrixOrthogonalize.3
  28-sep-2007:  Modified QccMatrixSVD() to work with matrics of any size

Version 0.55:
  29-aug-2007:  BETA release version 0.55
  03-jul-2007:  Added limits.h include
  17-may-2007:  Fixed num_symbols bug in QccENTHuffmanDesign()

QccPack-0.61-1/src/libQccPackVID/utils/memc.c
/*
 *
 * QccPack: Quantization, compression, and coding utilities
 * Copyright (C) 1997-2016  James E. Fowler
 * 
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */


#include "memc.h"


#define USG_STRING "[-fp %:] [-hp %:] [-qp %:] [-ep %:] [-f1 %: %s:filter1] [-f2 %: %s:filter2] [-f3 %: %s:filter3] [-w %d:window] [-b %d:block_size] [-mc %: %s:motion_compensated_frame] [-srf %: %s:subpixel_reference_frame] [-r %: %s:residual_frame] %s:reference_frame %s:current_frame %s:mvfile"

QccIMGImage ReferenceFrame;
QccIMGImage CurrentFrame;

QccString MVFilename;
int NumCols1, NumRows1;
int NumCols2, NumRows2;

int WindowSize = 15;
int BlockSize = 16;

QccIMGImageComponent HorizontalMotion;
QccIMGImageComponent VerticalMotion;

int FullPixelSpecified = 0;
int HalfPixelSpecified = 0;
int QuarterPixelSpecified = 0;
int EighthPixelSpecified = 0;
int MotionAccuracy = QCCVID_ME_FULLPIXEL;

QccIMGImageComponent ReferenceFrame2;

int FilterSpecified1 = 0;
QccString FilterFilename1;
QccFilter Filter1;
int FilterSpecified2 = 0;
QccString FilterFilename2;
QccFilter Filter2;
int FilterSpecified3 = 0;
QccString FilterFilename3;
QccFilter Filter3;

int CompensatedFrameSpecified = 0;
QccIMGImage CompensatedFrame;

int SubpixelReferenceFrameSpecified = 0;
QccIMGImage SubpixelReferenceFrame;

int ResidualFrameSpecified = 0;
QccIMGImage ResidualFrame;


int main(int argc, char *argv[])
{
  FILE *infile;

  QccInit(argc, argv);

  QccIMGImageInitialize(&ReferenceFrame);
  QccIMGImageInitialize(&CurrentFrame);
  QccIMGImageComponentInitialize(&HorizontalMotion);
  QccIMGImageComponentInitialize(&VerticalMotion);
  QccIMGImageComponentInitialize(&ReferenceFrame2);
  QccFilterInitialize(&Filter1);
  QccFilterInitialize(&Filter2);
  QccFilterInitialize(&Filter3);
  QccIMGImageInitialize(&CompensatedFrame);
  QccIMGImageInitialize(&SubpixelReferenceFrame);
  QccIMGImageInitialize(&ResidualFrame);

  if (QccParseParameters(argc, argv,
			 USG_STRING,
                         &FullPixelSpecified,
                         &HalfPixelSpecified,
                         &QuarterPixelSpecified,
                         &EighthPixelSpecified,
                         &FilterSpecified1,
                         FilterFilename1,
                         &FilterSpecified2,
                         FilterFilename2,
                         &FilterSpecified3,
                         FilterFilename3,
			 &WindowSize,
                         &BlockSize,
                         &CompensatedFrameSpecified,
                         CompensatedFrame.filename,
                         &SubpixelReferenceFrameSpecified,
                         SubpixelReferenceFrame.filename,
                         &ResidualFrameSpecified,
                         ResidualFrame.filename,
			 ReferenceFrame.filename,
			 CurrentFrame.filename,
                         MVFilename))
    QccErrorExit();
  
  if (EighthPixelSpecified)
    MotionAccuracy = QCCVID_ME_EIGHTHPIXEL;
  else
    if (QuarterPixelSpecified)
      MotionAccuracy = QCCVID_ME_QUARTERPIXEL;
    else
      if (HalfPixelSpecified)
        MotionAccuracy = QCCVID_ME_HALFPIXEL;
      else
        if (FullPixelSpecified)
          MotionAccuracy = QCCVID_ME_FULLPIXEL;
  
  if (FilterSpecified1)
    {
      if ((infile = QccFileOpen(FilterFilename1, "r")) == NULL)
        {
          QccErrorAddMessage("%s: Error calling QccFileOpen()",
                             argv[0]);
          QccErrorExit();
        }
      if (QccFilterRead(infile, &Filter1))
        {
          QccErrorAddMessage("%s: Error calling QccFilterRead()",
                             argv[0]);
          QccErrorExit();
        }
      QccFileClose(infile);
    }
  if (FilterSpecified2)
    {
      if ((infile = QccFileOpen(FilterFilename2, "r")) == NULL)
        {
          QccErrorAddMessage("%s: Error calling QccFileOpen()",
                             argv[0]);
          QccErrorExit();
        }
      if (QccFilterRead(infile, &Filter2))
        {
          QccErrorAddMessage("%s: Error calling QccFilterRead()",
                             argv[0]);
          QccErrorExit();
        }
      QccFileClose(infile);
    }
  if (FilterSpecified3)
    {
      if ((infile = QccFileOpen(FilterFilename3, "r")) == NULL)
        {
          QccErrorAddMessage("%s: Error calling QccFileOpen()",
                             argv[0]);
          QccErrorExit();
        }
      if (QccFilterRead(infile, &Filter3))
        {
          QccErrorAddMessage("%s: Error calling QccFilterRead()",
                             argv[0]);
          QccErrorExit();
        }
      QccFileClose(infile);
    }

  if (QccIMGImageRead(&ReferenceFrame))
    {
      QccErrorAddMessage("%s: Error calling QccIMGImageRead()",
			 argv[0]);
      QccErrorExit();
    }
  if (QccIMGImageRead(&CurrentFrame))
    {
      QccErrorAddMessage("%s: Error calling QccIMGImageRead()",
			 argv[0]);
      QccErrorExit();
    }

  if (QccIMGImageColor(&ReferenceFrame) || QccIMGImageColor(&CurrentFrame))
    {
      QccErrorAddMessage("%s: Both images must be grayscale",
			 argv[0]);
      QccErrorExit();
    }

  if (QccIMGImageGetSize(&ReferenceFrame,
                         &NumRows1, &NumCols1))
    {
      QccErrorAddMessage("%s: Error calling QccIMGImageGetSize()",
			 argv[0]);
      QccErrorExit();
    }
  if (QccIMGImageGetSize(&CurrentFrame,
                         &NumRows2, &NumCols2))
    {
      QccErrorAddMessage("%s: Error calling QccIMGImageGetSize()",
			 argv[0]);
      QccErrorExit();
    }

  if ((NumRows1 % BlockSize) || (NumCols1 % BlockSize))
    {
      QccErrorAddMessage("%s: Image dimensions must be integer multiple of block size %d",
			 argv[0],
			 BlockSize);
      QccErrorExit();
    }

  if ((NumRows1 != NumRows2) || (NumCols1 != NumCols2))
    {
      QccErrorAddMessage("%s: Both images must be same size",
			 argv[0]);
      QccErrorExit();
    }

  HorizontalMotion.num_rows = NumRows1 / BlockSize;
  HorizontalMotion.num_cols = NumCols1 / BlockSize;
  if (QccIMGImageComponentAlloc(&HorizontalMotion))
    {
      QccErrorAddMessage("%s: Error calling QccIMGImageComponentAlloc()",
			 argv[0]);
      QccErrorExit();
    }
  VerticalMotion.num_rows = NumRows1 / BlockSize;
  VerticalMotion.num_cols = NumCols1 / BlockSize;
  if (QccIMGImageComponentAlloc(&VerticalMotion))
    {
      QccErrorAddMessage("%s: Error calling QccIMGImageComponentAlloc()",
			 argv[0]);
      QccErrorExit();
    }

  if (WindowSize <= 0)
    {
      QccErrorAddMessage("%s: Window size must be positive, non-zero",
			 argv[0]);
      QccErrorExit();
    }

  switch (MotionAccuracy)
    {
    case QCCVID_ME_FULLPIXEL:
      ReferenceFrame2.num_rows = ReferenceFrame.Y.num_rows;
      ReferenceFrame2.num_cols = ReferenceFrame.Y.num_cols;
      break;
    case QCCVID_ME_HALFPIXEL:
      ReferenceFrame2.num_rows = 2 * ReferenceFrame.Y.num_rows;
      ReferenceFrame2.num_cols = 2 * ReferenceFrame.Y.num_cols;
      break;
    case QCCVID_ME_QUARTERPIXEL:
      ReferenceFrame2.num_rows = 4 * ReferenceFrame.Y.num_rows;
      ReferenceFrame2.num_cols = 4 * ReferenceFrame.Y.num_cols;
      break;
    case QCCVID_ME_EIGHTHPIXEL:
      ReferenceFrame2.num_rows = 8 * ReferenceFrame.Y.num_rows;
      ReferenceFrame2.num_cols = 8 * ReferenceFrame.Y.num_cols;
      break;
    }

  if (QccIMGImageComponentAlloc(&ReferenceFrame2))
    {
      QccErrorAddMessage("%s: Error calling QccIMGImageComponentAlloc()",
			 argv[0]);
      QccErrorExit();
    }

  if (QccVIDMotionEstimationCreateReferenceFrame(&ReferenceFrame.Y,
                                                 &ReferenceFrame2,
                                                 MotionAccuracy,
                                                 ((FilterSpecified1) ?
                                                  &Filter1 : NULL),
                                                 ((FilterSpecified2) ?
                                                  &Filter2 : NULL),
                                                 ((FilterSpecified3) ?
                                                  &Filter3 : NULL)))
    {
      QccErrorAddMessage("%s: Error calling QccVIDMotionEstimationCreateReferenceFrame()",
			 argv[0]);
      QccErrorExit();
    }

  if (QccVIDMotionEstimationFullSearch(&CurrentFrame.Y,
                                       &ReferenceFrame2,
                                       &HorizontalMotion,
                                       &VerticalMotion,
                                       BlockSize,
                                       WindowSize,
                                       MotionAccuracy))
    {
      QccErrorAddMessage("%s: Error calling QccVIDMotionEstimationFullSearch()",
			 argv[0]);
      QccErrorExit();
    }

  if (QccVIDMotionVectorsWriteFile(&HorizontalMotion,
                                   &VerticalMotion,
                                   MVFilename,
                                   0))
    {
      QccErrorAddMessage("%s: Error calling QccVIDMotionVectorWriteFile()",
			 argv[0]);
      QccErrorExit();
    }

  if (CompensatedFrameSpecified || ResidualFrameSpecified)
    {
      if (QccIMGImageSetSize(&CompensatedFrame,
                             NumRows1,
                             NumCols1))
        {
          QccErrorAddMessage("%s: Error calling QccIMGImageSetSize()",
                             argv[0]);
          QccErrorExit();
        }
      if (QccIMGImageAlloc(&CompensatedFrame))
        {
          QccErrorAddMessage("%s: Error calling QccIMGImageAlloc()",
                             argv[0]);
          QccErrorExit();
        }
      
      if (QccVIDMotionEstimationCreateCompensatedFrame(&CompensatedFrame.Y,
                                                       &ReferenceFrame2,
                                                       &HorizontalMotion,
                                                       &VerticalMotion,
                                                       BlockSize,
                                                       MotionAccuracy))
        {
          QccErrorAddMessage("%s: Error calling QccVIDMotionEstimationCreateCompensatedFrame()",
                             argv[0]);
          QccErrorExit();
        }
    }
                                                       
  if (CompensatedFrameSpecified)
    {
      if (QccIMGImageDetermineType(&CompensatedFrame))
        {
          QccErrorAddMessage("%s: Error calling QccIMGImageDetermineType()",
                             argv[0]);
          QccErrorExit();
        }
      if (QccIMGImageColor(&CompensatedFrame))
        {
          QccErrorAddMessage("%s: Motion-compensated frame must be grayscale",
                             argv[0]);
          QccErrorExit();
        }
      
      if (QccIMGImageWrite(&CompensatedFrame))
        {
          QccErrorAddMessage("%s: Error calling QccIMGImageWrite()",
                             argv[0]);
          QccErrorExit();
        }
    }
  
  if (ResidualFrameSpecified)
    {
      if (QccIMGImageDetermineType(&ResidualFrame))
        {
          QccErrorAddMessage("%s: Error calling QccIMGImageDetermineType()",
                             argv[0]);
          QccErrorExit();
        }
      if (QccIMGImageColor(&ResidualFrame))
        {
          QccErrorAddMessage("%s: Motion-compensated frame must be grayscale",
                             argv[0]);
          QccErrorExit();
        }

      if (QccIMGImageSetSize(&ResidualFrame,
                             NumRows1,
                             NumCols1))
        {
          QccErrorAddMessage("%s: Error calling QccIMGImageSetSize()",
                             argv[0]);
          QccErrorExit();
        }
      if (QccIMGImageAlloc(&ResidualFrame))
        {
          QccErrorAddMessage("%s: Error calling QccIMGImageAlloc()",
                             argv[0]);
          QccErrorExit();
        }

      if (QccIMGImageComponentSubtract(&CurrentFrame.Y,
                                       &CompensatedFrame.Y,
                                       &ResidualFrame.Y))
        {
          QccErrorAddMessage("%s: Error calling QccIMGImageComponentSubtract()",
                             argv[0]);
          QccErrorExit();
        }
      if (QccIMGImageComponentAbsoluteValue(&ResidualFrame.Y))
        {
          QccErrorAddMessage("%s: Error calling QccIMGImageComponentAbsoluteValue()",
                             argv[0]);
          QccErrorExit();
        }

      if (QccIMGImageWrite(&ResidualFrame))
        {
          QccErrorAddMessage("%s: Error calling QccIMGImageWrite()",
                             argv[0]);
          QccErrorExit();
        }
    }

  if (SubpixelReferenceFrameSpecified)
    {
      if (QccIMGImageDetermineType(&SubpixelReferenceFrame))
        {
          QccErrorAddMessage("%s: Error calling QccIMGImageDetermineType()",
                             argv[0]);
          QccErrorExit();
        }
      if (QccIMGImageColor(&SubpixelReferenceFrame))
        {
          QccErrorAddMessage("%s: Motion-compensated frame must be grayscale",
                             argv[0]);
          QccErrorExit();
        }
      
      if (QccIMGImageSetSize(&SubpixelReferenceFrame,
                             ReferenceFrame2.num_rows,
                             ReferenceFrame2.num_cols))
        {
          QccErrorAddMessage("%s: Error calling QccIMGImageSetSize()",
                             argv[0]);
          QccErrorExit();
        }
      SubpixelReferenceFrame.Y.image = ReferenceFrame2.image;
      
      if (QccIMGImageWrite(&SubpixelReferenceFrame))
        {
          QccErrorAddMessage("%s: Error calling QccIMGImageWrite()",
                             argv[0]);
          QccErrorExit();
        }
    }
  
  QccIMGImageFree(&ReferenceFrame);
  QccIMGImageFree(&CurrentFrame);
  QccIMGImageComponentFree(&HorizontalMotion);
  QccIMGImageComponentFree(&VerticalMotion);
  QccIMGImageComponentFree(&ReferenceFrame2);
  QccFilterFree(&Filter1);
  QccFilterFree(&Filter2);
  QccFilterFree(&Filter3);
  QccIMGImageFree(&CompensatedFrame);

  QccExit;
}
